products/batch_exports/backend/temporal/record_batch_model.py

```diff
@@ -0,0 +1,212 @@
+                ),
+            ]
+        )
+        if data_interval_start is not None:
+            where_and.exprs.extend(
+                [
+                    ast.CompareOperation(
+                        op=ast.CompareOperationOp.GtEq,
+                        left=ast.Field(chain=["_inserted_at"]),
+                        right=ast.Constant(value=data_interval_start),
+                    ),
+                    # include $end_timestamp because hogql uses this to add a where clause to the inner query
+                    ast.CompareOperation(
+                        op=ast.CompareOperationOp.GtEq,
+                        left=ast.Field(chain=["$end_timestamp"]),
+                        right=ast.Constant(value=data_interval_start),
+                    ),
+                ]
+            )
+
+        hogql_query.where = where_and
+
+        return hogql_query
+
+    async def as_query_with_parameters(
+        self, data_interval_start: dt.datetime | None, data_interval_end: dt.datetime
+    ) -> tuple[Query, QueryParameters]:
+        """Produce a printed query and any necessary ClickHouse query parameters."""
+        hogql_query = self.get_hogql_query(data_interval_start, data_interval_end)
+        context = await self.get_hogql_context(self.team_id)
+
+        prepared_hogql_query = await database_sync_to_async(prepare_ast_for_printing)(
+            hogql_query, context=context, dialect="clickhouse", stack=[]
+        )
+        assert prepared_hogql_query is not None
+        context.output_format = "ArrowStream"
+        printed = print_prepared_ast(
+            prepared_hogql_query,
+            context=context,
+            dialect="clickhouse",
+            stack=[],
+        )
+        return printed, context.values
+
+    async def as_insert_into_s3_query_with_parameters(
+        self,
+        data_interval_start: dt.datetime | None,
+        data_interval_end: dt.datetime,
+        s3_folder: str,
+        s3_key: str,
+        s3_secret: str,
+        num_partitions: int,
+    ) -> tuple[Query, QueryParameters]:
+        """Produce a printed query and any necessary ClickHouse query parameters."""
+        hogql_query = self.get_hogql_query(data_interval_start, data_interval_end)
+        context = await self.get_hogql_context(self.team_id)
+
+        prepared_hogql_query = await database_sync_to_async(prepare_ast_for_printing)(
+            hogql_query, context=context, dialect="clickhouse", stack=[]
+        )
+        assert prepared_hogql_query is not None
+        printed = print_prepared_ast(
+            prepared_hogql_query,
+            context=context,
+            dialect="clickhouse",
+            stack=[],
+        )
+        insert_query = f"""
+INSERT INTO FUNCTION
+   s3(
+       '{s3_folder}/export_{{{{_partition_id}}}}.arrow',
+       '{s3_key}',
+       '{s3_secret}',
+       'ArrowStream'
+    )
+    PARTITION BY rand() %% {num_partitions}
+{printed}
+"""
```

Comment:
Yeah, I don't really think we can do inserts with hogql, this is likely as good as it gets.